### PR TITLE
unsupervised ml model sample notebook

### PR DESCRIPTION
**Pull request by moonlanderr**
\<insert pull request description here\>

-----

# Checklist

Please go through each entry in the below checklist and mark an 'X' if that condition has been met. Every entry should be marked with an 'X' to be get the Pull Request approved.

- [ ] All `import`s are in the first cell? First block of imports are standard libraries, second block are 3rd party libraries, third block are all `arcgis` imports? Note that in some cases, for samples, it is a good idea to keep the imports next to where they are used, particularly for uncommonly used features that we want to highlight.
- [ ] All `GIS` object instantiations are one of the following?
    - `gis = GIS()`
    - `gis = GIS('https://www.arcgis.com', 'arcgis_python', 'P@ssword123')`
    - `gis = GIS(profile="your_online_profile")`
    - `gis = GIS('https://pythonapi.playground.esri.com/portal', 'arcgis_python', 'amazing_arcgis_123')`
    - `gis = GIS(profile="your_enterprise_portal")`
- [ ] If this notebook requires setup or teardown, did you add the appropriate code to `./misc/setup.py` and/or `./misc/teardown.py`?
- [ ] If this notebook references any portal items that need to be staged on AGOL/Python API playground, did you coordinate with a Python API team member to stage the item the correct way with the api\_data\_owner user?
- [ ] Code refactored & split out across multiple cells, useful comments?
- [ ] Consistent voice/tense/narrative style? Thoroughly checked for typos?
- [ ] All images used like `` instead of ``? All map widgets contain a static image preview? (Call `mapview_inst.take_screenshot()` to do so)
- [ ] All file paths are constructed in an OS-agnostic fashion with `os.path.join()`? (Instead of `r"\foo\bar"`, `os.path.join(os.path.sep, "foo", "bar")`, etc.)
- [ ] **IF YOU WANT THIS SAMPLE TO BE DISPLAYED ON THE DEVELOPERS.ARCGIS.COM WEBSITE**, ping @ DavidJVitale so he can add it to the list for the next deploy

### CODE REVIEW

**Comment by priyankatuteja** [comment]
@KarthikDutt Please help in review of this notebook.

File: samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb [hunk 1]
```diff
@@ -0,0 +1,1169 @@
+{
```
Comment (priyankatuteja) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>please be careful with credentials while sending pr. also provide 'home' for arcgis online and credentials for enterprise pythonapi playground </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (priyankatuteja) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>The raster should exist on pythonapi playground portal</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (priyankatuteja) [reply samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: @rhea-j I would suggest changing the password for your portal asap

Comment (moonlanderr) [reply samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: credential changed

Comment (moonlanderr) [reply samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: raster added to playground portal

**Comment by priyankatuteja** [comment]
@moonlanderr I just ran this notebook and found that it breaks at 2 cells.

`pred_new = model.predict(explanatory_rasters=[raster_amazon_13bands], prediction_type='raster',output_layer_name='deforest_predicted2',output_raster_path=r"/tmp/result4.tif")` # change the path to /tmp/result4.tif

`# publish the scene to the portal
amazon_scene.save('amazon_scene_july9', gis=gis_enterp)`

It breaks with name already exists, add dt.now(...) to create a unique name

**Comment by priyankatuteja** [comment]
@moonlanderr I just ran this notebook and found that it breaks at 2 cells.

`pred_new = model.predict(explanatory_rasters=[raster_amazon_13bands], prediction_type='raster',output_layer_name='deforest_predicted2',output_raster_path=r"/tmp/result4.tif")` # change the path to /tmp/result4.tif

`# publish the scene to the portal
amazon_scene.save('amazon_scene_july9', gis=gis_enterp)`

It breaks with name already exists, add dt.now(...) to create a unique name

**Comment by moonlanderr** [comment]
> @moonlanderr I just ran this notebook and found that it breaks at 2 cells.
> 
> `pred_new = model.predict(explanatory_rasters=[raster_amazon_13bands], prediction_type='raster',output_layer_name='deforest_predicted2',output_raster_path=r"/tmp/result4.tif")` # change the path to /tmp/result4.tif
> 
> `# publish the scene to the portal amazon_scene.save('amazon_scene_july9', gis=gis_enterp)`
> 
> It breaks with name already exists, add dt.now(...) to create a unique name

the new path is changed such that it will create the layer in the current folder of the working directory of the notebook

**Review by priyankatuteja — APPROVED** [review]

**Comment by priyankatuteja** [comment]
@BP-Ent Could you please help in doc review

File: samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb [hunk 2]
```diff
@@ -0,0 +1,1190 @@
...
```
Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Deforestation around the world has reached a critical level, causing irreversible damage to environmental sustainability that is contributing to climate change around the world. Widespread forest fires, from the Amazon Basin in Brazil, to the west coast of the United States, are raging all year-round. This notebook will allow us to detect deforested areas in the Brazilian Amazon rainforest, using satellite imagery.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Here, we use Sentinel-2 imagery, which has a high resolution of 10m and 13 bands. This imagery is accessed from the ArcGIS Enterprise portal, where it is sourced from the AWS collection. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>The area of interest is defined using the four latitude and longitude values from a certain region of the Amazon rainforest where a considerable area of forest has been deforested, as can be seen from the images above.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Here, we select all of the scenes from the sentinel imagery containing the area of interest for our study.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>In the above scene, the brown patches are the deforested areas that are to be identified.  This selected scene is then published to the portal.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>The first part of model building consists of defining the preprocessors, which will be used to scale the bands before feeding them into the model. The band names use the conventional naming method of the imagery name with an id number appended at the end as follows:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Sentinel-2 imagery has 13 bands, of which 4 bands, namely the blue, green, red, and near infrared bands, we will use here for modelling. These bands work well for differentiating green forested areas from barren land. The band information, along with the band name and their respective ids, are obtained for selecting the required bands.     </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Here, the imagery name is 'Hosted/amazon_scene_april9'. Similarly, the names of the blue, green, red, and near infrared bands would be 'Hosted/amazon_scene_april9_1', 'Hosted/amazon_scene_april9_2', 'Hosted/amazon_scene_april9_3', 'Hosted/amazon_scene_april9_7' respectively. These bands will be used for defining the preprocessors.   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Here, in the explanatory raster, we pass the name of the explanatory raster and the selected bands by their id's — 1 for blue, 2 for green, 3 for red, and 7 for NIR, as follows:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='32'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Once the data is prepared, an unsupervised model of k-means clustering from scikit-learn is used for clustering the pixels into deforested areas and forested areas. The clustering model is passed inside MLModel as follows, with the number of clusters set as three in the parameters.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Now the model is ready to be trained, and will label the pixels as being one of three classes, either forested, slightly deforested, or highly deforested. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>Next, the trained model is used to predict clusters within the entire selected scene of the Amazon rainforest. This is passed as the explanatory raster, with the prediction type as raster and a local path provided for output. The <code>output_layer_name</code> parameter can also be used for publishing. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>The resulting raster with the predicted classes of deforested areas and forested areas is now read back for visualization. The predicted local raster is then published in the portal and can be accessed here. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>The model has correctly labelled the deforested areas in white, distinguishing them from the rest of the forested areas in black.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (BP-Ent) [inline samples/04_gis_analysts_data_scientists/detecting-deforestation-using-kmeans-clustering-on-sentinel-imagery.ipynb:1]: <p>In this sample notebook we were able to detect deforestation in the Amazon rainforest using the unsupervised model of k-means clustering on satellite imagery. This was implemented via the MLModel framework, which exhibited the application of several unsupervised models from the scikit learn library on raster imagery.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Esri/arcgis-python-api/pull/975/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='50'/>

**Review by BP-Ent — CHANGES_REQUESTED** [review]
Suggested changes made on reviewnb

**Comment by moonlanderr** [comment]
> Suggested changes made on reviewnb

@BP-Ent , suggestions added please check

**Comment by moonlanderr** [comment]
@BP-Ent please let me know if you have any update

**Comment by BP-Ent** [comment]
@moonlanderr Looks good!

**Comment by moonlanderr** [comment]
@BP-Ent thank you!

**Comment by priyankatuteja** [comment]
@rohitgeo @AtmaMani This notebook can be merged

**Review by rohitgeo — APPROVED** [review]

**Pull request merged**